src/osd.c

```diff
@@ -223,6 +223,138 @@ preview_cycled_view(struct view *view)
+			OSD_ITEM_WIDTH, OSD_ITEM_HEIGHT);
+		cairo_stroke(cairo);
+	}
+
+	/* Set up text rendering */
+	set_cairo_color(cairo, theme->osd_label_text_color);
+	PangoLayout *layout = pango_cairo_create_layout(cairo);
+	pango_layout_set_width(layout,
+		(OSD_ITEM_WIDTH - 2 * OSD_ITEM_PADDING) * PANGO_SCALE);
+	pango_layout_set_ellipsize(layout, PANGO_ELLIPSIZE_END);
+
+	PangoFontDescription *desc = font_to_pango_desc(&rc.font_osd);
+	pango_layout_set_font_description(layout, desc);
+
+	PangoTabArray *tabs = pango_tab_array_new_with_positions(2, TRUE,
+		PANGO_TAB_LEFT, OSD_TAB1, PANGO_TAB_LEFT, OSD_TAB2);
+	pango_layout_set_tabs(layout, tabs);
+	pango_tab_array_free(tabs);
+
+	pango_cairo_update_layout(cairo, layout);
+
+	int y = OSD_BORDER_WIDTH;
+
+	/* Center text entries on the y axis */
+	int y_offset = (OSD_ITEM_HEIGHT - font_height(&rc.font_osd)) / 2;
+	y += y_offset;
+
+	/* Draw workspace indicator */
+	if (show_workspace) {
+		/* Center workspace indicator on the x axis */
+		int x = font_width(&rc.font_osd, workspace_name);
+		x = (OSD_ITEM_WIDTH - x) / 2;
+		cairo_move_to(cairo, x, y);
+		PangoWeight weight = pango_font_description_get_weight(desc);
+		pango_font_description_set_weight(desc, PANGO_WEIGHT_BOLD);
+		pango_layout_set_font_description(layout, desc);
+		pango_layout_set_text(layout, workspace_name, -1);
+		pango_cairo_show_layout(cairo, layout);
+		pango_font_description_set_weight(desc, weight);
+		pango_layout_set_font_description(layout, desc);
+		y += OSD_ITEM_HEIGHT;
+	}
+	pango_font_description_free(desc);
+
+	struct buf buf;
+	buf_init(&buf);
+
+	/* Draw text for each node */
+	wl_list_for_each_reverse(node, node_list, link) {
+		struct view *view = node_view_from_node(node);
+		if (!isfocusable(view)) {
+			continue;
+		}
+		buf.len = 0;
+		cairo_move_to(cairo, OSD_BORDER_WIDTH + OSD_ITEM_PADDING, y);
+
+		switch (view->type) {
+		case LAB_XDG_SHELL_VIEW:
+			buf_add(&buf, "[xdg-shell]\t");
+			buf_add(&buf, get_formatted_app_id(view));
+			buf_add(&buf, "\t");
+			break;
+#if HAVE_XWAYLAND
+		case LAB_XWAYLAND_VIEW:
+			buf_add(&buf, "[xwayland]\t");
+			buf_add(&buf, view_get_string_prop(view, "class"));
+			buf_add(&buf, "\t");
+			break;
+#endif
+		}
+
+		if (is_title_different(view)) {
+			buf_add(&buf, view_get_string_prop(view, "title"));
+		}
+
+		pango_layout_set_text(layout, buf.buf, -1);
+		pango_cairo_show_layout(cairo, layout);
+		y += OSD_ITEM_HEIGHT;
+	}
+	free(buf.buf);
+	g_object_unref(layout);
+
+	cairo_surface_flush(surf);
+}
+
+/* returns index of node with the given view, or -1 if not found */
+static int
+find_view_index(struct wl_list *node_list, struct view *view_to_find)
+{
+	struct wlr_scene_node *node;
+	int i = 0;
+	wl_list_for_each_reverse(node, node_list, link) {
+		struct view *view = node_view_from_node(node);
+		if (!isfocusable(view)) {
+			continue;
+		}
+		if (view == view_to_find) {
+			return i;
+		}
+		i += 1;
+	}
+	return -1;
+}
+
 void
 osd_update(struct server *server)
```

Comment:
This function can be cleared up completely by moving the parts relevant to the OSD to the new function.
All that is left then looks something like this:
```c
void
osd_update(struct server *server)
{
	if (wl_list_empty(&server->workspace_current->tree->children)
			|| !server->osd_state.cycle_view) {
		osd_finish(server);
		return;
	}

	/* Render the actual OSD */
	struct output *output;
	wl_list_for_each(output, &server->outputs, link) {
		destroy_osd_nodes(output);
		if (!output->wlr_output->enabled) {
			continue;
		}
		render_osd(output);
	}

	if (rc.cycle_preview_outlines) {
		osd_update_preview_outlines(server->osd_state.cycle_view);
	}

	if (rc.cycle_preview_contents) {
		preview_cycled_view(server->osd_state.cycle_view);
	}
}
```

The whole output loop can then be further enclosed in a new condition like `if (rc.cycle_osd) { .. }` (in a separate commit / PR).